File: SDE sheet/day2/mergeSORT.cpp
#include<iostream>
using namespace std;


// void display(int a[],int n)
// {
// 	cout<<"the sorted array is ->";
// 	for (int i = 0; i < n; i++)
// 	{
// 		cout<<a[i]<<" , ";
// 	}
// 	return ;
// }

void merge(int arr[],int l ,int r,int mid)
{
	int i = l;
	int j=mid+1;
	int k=0;

	int temp[r-l+1];
	while(i<=mid && j<=r)
	{
		if(arr[i] <= arr[j])
		{
			temp[k]=arr[i];
			i++;
			k++;
		}
		else
		{
			temp[k]=arr[j];
			j++;
			k++;
		}
	}
	while(i<=mid)
	{
			temp[k]=arr[i];
			i++;
			k++;
	}
	while(j<=r)
	{
			temp[k]=arr[j];
			j++;
			k++;
	}
	for (int i = l; i <= r; i++)
	{
		arr[i]=temp[i-l];
	}
}

void mergesort(int arr[],int l,int r)
{
	if(l<r)
	{
		int mid=(l+r)/2;
		mergesort(arr,l,mid);
		mergesort(arr,mid+1,r);
		merge(arr,l,r,mid);
	}
}

int main()
{
	int arr[]={8,3,4,1,7,6,5,0};
	int l=0 ,r =7;
	mergesort(arr,l,r);

	// display(arr,8);

	for (int i = 0; i < 8; ++i)
	{
		/* code */cout<<arr[i]<<" ";
	}
}

File: java-bootcamp/recursion/src/SortingUsingRecursion/mergeSort.java
package SortingUsingRecursion;

import java.util.Arrays;

public class mergeSort {
    public static void main(String[] args) {

        int arr[]={5,4,3,2,1};
//        System.out.println(Arrays.toString(mergeSort(arr)));
//        int arr1[]={9,8,7,6,5,4,3};
        mergeSortinPlace(arr,0,arr.length);
        System.out.println(Arrays.toString(arr));
    }

    static int[] mergeSort(int[] arr)
    {
        if(arr.length==1)
        {
            return arr;
        }

        int middle= arr.length/2;
        int[] left=mergeSort(Arrays.copyOfRange(arr,0,middle));
        int[] right= mergeSort(Arrays.copyOfRange(arr,middle,arr.length));

        return merge(right,left);

    }

    static int[] merge(int[] first,int[] second)
    {
        int[] mix= new int[first.length + second.length ];
        int i=0;
        int j=0;
        int k=0;

        while(i < first.length && j < second.length)
        {
            if(first[i] < second[j])
            {
                mix[k]=first[i];
                i++;
            }
            else
            {
                mix[k]= second[j];
                j++;
            }
            k++;
        }


//        coping the rest of the array
        while(i< first.length)

        {
            mix[k]=first[i];
            i++;k++;
        }
        while(j<second.length)
        {
            mix[k]=second[j];
            j++;k++;
        }
        return mix;
    }

    static void mergeSortinPlace(int[] arr,int s,int e)
    {
        if(e - s == 1) {
            return;
        }
        int mid= (s+e)/2;

        mergeSortinPlace(arr,s,mid);
        mergeSortinPlace(arr,mid,e);

        mergeInPlace(arr,s,mid,e);
    }
    static void mergeInPlace(int[] arr,int s,int m,int e)
    {
        int mix[]= new int[e-s];

        int i=s;
        int j=m;
        int k=0;


        while( i < m && j < e )
        {
            if(arr[i]<arr[j])
            {
                mix[k]=arr[i];
                i++;
            }
            else
            {
                mix[k]=arr[j];
                j++;
            }
            k++;
        }
        //coping the rst of the array

        while( i < m )
        {
            mix[k]=arr[i];
            i++;
            k++;
        }
        while(j<e)
        {
            mix[k]=arr[j];
            j++;
            k++;
        }
//      coping the array mix to the original array
        for (int l = 0 ; l < mix.length; l++) {
            arr[s+l]=mix[l];
        }
    }
}
